cmd/registry/controller/parser.go

```diff
@@ -78,39 +110,51 @@ func extendDependencyPattern(
 
 }
 
+// This function is used in case of receipt artifacts where the name of the target resource to create is not known
 func resourceNameFromGroupKey(
 	resourcePattern string,
 	groupKey string) (string, error) {
-	// Derives the resource name from the provided resourcePattern and dependencyName.
+	// Derives the resource name from the provided resourcePattern and groupKey.
 	// Example:
 	// 1) resourcePattern: projects/demo/locations/global/apis/-/versions/-/specs/-
-	//    dependencyName: projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/complexity
+	//    groupKey: projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/complexity
 	//    returns projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml
 	// 2) resourcePattern: projects/demo/locations/global/apis/petstore/versions/-/specs/-/artifacts/custom-artifact
-	//    dependencyName: projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/complexity
+	//    groupKey: projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/complexity
 	//    returns projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/custom-artifact
 
-	// Replace apis/- pattern with the corresponding name and so on.
-	// dependency.GetApi() returns the full api name projects/demo/locations/global/apis/petstore
-	// We use stringsSplit()[-1] to extract only the API name
-	// apiPattern := regexp.MustCompile(`/apis/-`)
-	apiName := strings.Split(extractEntityName(groupKey, "api"), "/")
-	resourceName := strings.ReplaceAll(resourcePattern, "/apis/-",
-		fmt.Sprintf("/apis/%s", apiName[len(apiName)-1]))
-
-	versionName := strings.Split(extractEntityName(groupKey, "version"), "/")
-	resourceName = strings.ReplaceAll(resourceName, "/versions/-",
-		fmt.Sprintf("/versions/%s", versionName[len(versionName)-1]))
+	groupName, err := parsePattern(groupKey)
```

Comment:
groupKeys can only be resourceNames and not collections.  You can checkout the `getGroupKey()` function on line 232, which generates these keys. `parsePattern` is a common function which is reused here.

Reply:
Okay. If we ever accidentally call this function with a collection instead of a resource name as expected, the function will return without error. If it's easy to validate that would be a good defensive measure.

Reply:
Done. Broke the `parsePattern()` into separate functions which individually parse Collections ane Names.